exploration/0003-formatted-parts.md

```diff
@@ -0,0 +1,187 @@
+
+## Background
+
+Past examples have shown us that if we don't provide a formatter to parts,
+the string output will be re-parsed and re-processed by users.
+
+## Use-Cases
+
+- Markup elements
+- Non-string values
+- Message post-processors
+
+## Requirements
+
+- Define an iterable sequence of formatted part objects.
+- Include metadata for each part, such as type, source, direction, and locale.
+- Allow the representation of non-string values.
+- Allow the representation of values that consist of an iterable sequence of formatted parts.
+- Be able to represent each resolved value of a pattern with any number of formatted parts, including none.
+- Define the formatted parts in a manner that allows synonymous but appropriate implementations in different programming languages.
+
+## Constraints
+
+- The JS Intl formatters already include formatted-parts representations for each supported data type.
+  The JS implementation of the MF2 formatted-parts representation should be able to match their structure,
+  at least as far as that's possible and appropriate.
+
+## Proposed Design
+
+The formatted-parts API is included in the spec as an optional but recommended formatting target.
+
+The shape of the formatted-parts output is defined in a manner similar to the data model,
+which includes TypeScript, JSON Schema, and XML DTD definitions of the same data structure.
+
+At the top level, the formatted-parts result is an iterable sequence of parts.
+Parts corresponding to each _text_ can be simpler than those of _expressions_,
+as they do not have a `source` other than their `value`,
+or set any of the other possible metadata fields.
+
+```ts
+type MessageParts = Iterable<
+  MessageTextPart | MessageExpressionPart | MessageBiDiIsolationPart
+>;
+
+interface MessageTextPart {
+  type: "text";
+  value: string;
+}
+```
+
+For MessageExpressionPart, the `source` corresponds to the expression's fallback value.
+The parts' `dir` and `locale` may be defined if overridden by an expression attribute or set otherwise.
+Each part should have at most one of `value` or `parts` defined;
+some may have none.
+
+```ts
+interface MessageExpressionPart {
+  type: string;
+  source: string;
+  parts?: Iterable<{ type: string; value: unknown }>;
+  value?: unknown;
+  dir?: "ltr" | "rtl" | "auto";
+  locale?: string;
+}
+```
+
+The bidi isolation strategies included in the spec may require
+the insertion of MessageBiDiIsolationParts in the formatted-parts output.
+
+```ts
+interface MessageBiDiIsolationPart {
+  type: "bidiIsolation";
+  value: "\u2066" | "\u2067" | "\u2068" | "\u2069"; // LRI | RLI | FSI | PDI
+}
+```
+
+Some of the MessageExpressionPart instances may be further defined
+without reference to the function registry.
+
+Unannotated expressions with a _literal_ operand
+are represented by MessageStringPart.
+As with MessageTextPart,
+the `value` of MessageStringPart is always a string.
```

Comment:
Is there a need for the distinction? It is simply the source?

Reply:
In addition to `source` the `dir` and `locale` are not present on MessageTextPart.

Reply:
Why don't text parts have a direction and language? Are they inheriting from the message?

Aside: we don't necessarily want to span every section of text, that is, this is better:

```html
<span lang=foo dir=ltr>blah blah blah 
   <span lang=bar dir=auto>placeholder here</span> blah blah</span>
```

than this:

```html
<span lang=foo dir=ltr>blah blah blah </span>
   <span lang=bar dir=auto>placeholder here</span><span lang=foo dir=ltr> blah blah</span>
```

... which is why text parts would inherit language and base paragraph direction from the message.

Reply:
Text parts do inherit the language and direction from the message, so they'll be whatever the formatter was called with. But they're not necessary to be repeated on every part, because the caller already knows what they asked for, and they may well be formatting the message in a context with e.g. `<body lang=foo dir=ltr>` setting them at a much higher level.